### ISSUE TITLE
An enum with one variant that directly contains PhantomData should be FFI-safe.

### ISSUE THREAD

**Issue by shepmaster** [rust-lang/rust#107854]
I tried this code:

```rust
use core::marker::PhantomData;

#[repr(C)]
enum TheEnum {
    VariantWithPhantomData(u8, PhantomData<String>),
//    VariantWithStruct(StructWithPhantomData),
}

#[repr(C)]
struct StructWithPhantomData(u8, PhantomData<String>);

extern "C" fn demo(_x: TheEnum) {}
```

I expected to see no warnings about `TheEnum`.

Instead, this happened: 

```
warning: `extern` fn uses type `TheEnum`, which is not FFI-safe
  --> src/lib.rs:12:24
   |
12 | extern "C" fn demo(_x: TheEnum) {}
   |                        ^^^^^^^ not FFI-safe
   |
   = note: this enum contains a PhantomData field
note: the type is defined here
  --> src/lib.rs:4:1
   |
4  | enum TheEnum {
   | ^^^^^^^^^^^^
   = note: `#[warn(improper_ctypes_definitions)]` on by default
```

If I move the `PhantomData` to a struct (see `VariantWithPhantomData`), then the problem goes away. The problem goes away **even if I leave the original variant that it previously complained about**.

### Meta

This reproduces back to at least Rust 1.56 and to recent nightly versions:

`rustc --version --verbose`:
```
rustc 1.69.0-nightly (f3126500f 2023-02-02)
binary: rustc
commit-hash: f3126500f25114ba4e0ac3e76694dd45a22de56d
commit-date: 2023-02-02
host: aarch64-apple-darwin
release: 1.69.0-nightly
LLVM version: 15.0.7
```

**Comment by yyny** [comment]
Fixed in the latest release:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=4892487c437a52ab5e0fdec896d0b1e1
